test/e2e/utils_test.go

```diff
@@ -220,69 +221,111 @@ func getPreviousGitTag(path string, cur semver.Version) (semver.Version, error)
 }
 
 // getKongProxyIP takes a Service with Kong proxy ports and returns and its IP, or fails the test if it cannot.
-func getKongProxyIP(ctx context.Context, t *testing.T, env environments.Environment, svc *corev1.Service) string {
-	proxyIP := ""
-	require.NotEqual(t, svc.Spec.Type, svc.Spec.ClusterIP)
-	if svc.Spec.Type == corev1.ServiceTypeLoadBalancer {
+func getKongProxyIP(ctx context.Context, t *testing.T, env environments.Environment) string {
+	refreshService := func() *corev1.Service {
+		svc, err := env.Cluster().Client().CoreV1().Services(namespace).Get(ctx, "kong-proxy", metav1.GetOptions{})
+		require.NoError(t, err)
+		return svc
+	}
+
+	svc := refreshService()
+	require.NotEqual(t, svc.Spec.Type, corev1.ServiceTypeClusterIP, "ClusterIP service is not supported")
+
+	//nolint: exhaustive
+	switch svc.Spec.Type {
+	case corev1.ServiceTypeLoadBalancer:
+		return getKongProxyLoadBalancerIP(t, refreshService)
+	case corev1.ServiceTypeNodePort:
+		return getKongProxyNodePortIP(ctx, t, env, svc)
+	default:
+		t.Fatalf("unknown service type: %q", svc.Spec.Type)
+		return ""
+	}
+}
+
+func getKongProxyLoadBalancerIP(t *testing.T, refreshSvc func() *corev1.Service) string {
+	var resIP string
+	require.Eventually(t, func() bool {
+		svc := refreshSvc()
+
 		if len(svc.Status.LoadBalancer.Ingress) > 0 {
-			proxyIP = svc.Status.LoadBalancer.Ingress[0].IP
-			t.Logf("found loadbalancer IP for the Kong Proxy: %s", proxyIP)
+			ip := svc.Status.LoadBalancer.Ingress[0].IP
+			t.Logf("found loadbalancer IP for the Kong Proxy: %s", ip)
+			resIP = ip
+			return true
+		}
+
+		t.Log("no IP for LoadBalancer found yet")
+		return false
+	}, ingressWait, time.Second)
+
+	return resIP
+}
+
+func getKongProxyNodePortIP(ctx context.Context, t *testing.T, env environments.Environment, svc *corev1.Service) string {
+	var port corev1.ServicePort
+	for _, sport := range svc.Spec.Ports {
+		if sport.Name == "kong-proxy" || sport.Name == "proxy" {
+			port = sport
 		}
 	}
-	// the above failed to find an address. either the LB didn't provision or we're using a NodePort
-	if proxyIP == "" {
-		var port int32
-		for _, sport := range svc.Spec.Ports {
-			if sport.Name == "kong-proxy" || sport.Name == "proxy" {
-				port = sport.NodePort
+
+	// GKE clusters by default do not allow ingress traffic to its nodes
+	// TODO: consider adding an option to create firewall rules in KTF GKE provider
+	if env.Cluster().Type() == gke.GKEClusterType {
+		const kongProxyLocalPort = "9779"
```

Comment:
I sense potential for flakes when a port is occupied :) 

I wanted to suggest to use `0` which should theoretically pick a free port for you but since we're invoking `kubectl` we might not get that (unless we parse `kubectl`'s output).

🤔 Maybe we can try retrying this couple of times in case it fails due to an occupied port?

Reply:
How about this: bd35116? :)